### ISSUE TITLE
bug(core/lifelong): HAS_COMPLETED_INITIAL_TRAINING set to True on first training call in standard mode

### ISSUE THREAD

**Issue by g-k-s-03** [kubeedge/ianvs#572]
## Description

In `lifelong_learning.py`, the standard benchmark mode (`elif mode != 
'multi-inference'`) starts its loop at `r=1`, but `_train()` uses 
`rounds < 1` to determine whether the initial training has completed. 
Since `1 < 1` is always `False`, `HAS_COMPLETED_INITIAL_TRAINING` is 
set to `'True'` on the very first call — before any training has 
actually occurred.

## Files Involved

- `core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py`
  - Line 266: standard mode loop starts at `r=1`
  - Line 359: condition `if rounds < 1` is never `True` when called from standard mode

## Root Cause

The no-inference and hard-example-mining modes start their loops at 
`r=0`, so `rounds < 1` correctly evaluates to `True` on the first 
call and sets `HAS_COMPLETED_INITIAL_TRAINING = 'False'`. The standard 
mode starts at `r=1`, making the condition always `False`, so the env 
var is immediately set to `'True'` before the first training round.

## Impact

This is a silent correctness bug. On the very first training call, 
every lifelong learning algorithm receives 
`HAS_COMPLETED_INITIAL_TRAINING == 'True'` and assumes a prior model 
state exists. The algorithm either attempts to load a non-existent 
checkpoint or skips knowledge-base initialization. No exception is 
raised — training proceeds but builds on an incorrect initial state, 
silently corrupting all subsequent benchmark metrics (BWT, FWT, Matrix).

## Steps to Reproduce

Add the following assertion to any lifelong learning basemodel:

```python
def train(self, train_data):
    assert os.environ["HAS_COMPLETED_INITIAL_TRAINING"] == "False", \
        f"Got {os.environ['HAS_COMPLETED_INITIAL_TRAINING']} on first call"
```

Run any lifelong learning benchmark in standard mode. The assertion 
fails on the first round, confirming the env var is incorrectly set.

## Expected Behavior

`HAS_COMPLETED_INITIAL_TRAINING` should be `'False'` on the first 
training call and `'True'` on all subsequent calls, regardless of the 
loop start index.

## Proposed Fix

Change line 359 from:

```python
if rounds < 1:
```

to:

```python
if rounds <= 1:
```

This ensures the condition is `True` when `r=1` (first call in 
standard mode) and `False` for all subsequent rounds (`r=2,3,...`). 
The no-inference and hard-example-mining modes are unaffected since 
they call `_train` with `r=0`, which still satisfies `r <= 1`.

## Why This Is Multi-File

The bug is set in `lifelong_learning.py` but its impact is realized 
in user algorithm code (any `basemodel.py` consuming 
`HAS_COMPLETED_INITIAL_TRAINING`). The broken contract propagates 
silently from the paradigm layer into algorithm implementations, 
making it invisible unless you trace the env var across both layers.

**Comment by g-k-s-03** [comment]
I identified this bug and would like to work on the fix. @jaypume  @MooreZheng

**Comment by zjr060424-lab** [comment]
I reproduced the first-round state error described in this Issue against locked `main@37a9c60a9747af0cfe3170f84249bc349c56e8d5` and compared it with `PR #573@96522f6b14fcfd408f2f9d3c7e94fc1bbde829ce`.

## Source path

In standard mode, the current Core loop starts at `r = 1`:

- [`lifelong_learning.py:263–271`](https://github.com/kubeedge/ianvs/blob/37a9c60a9747af0cfe3170f84249bc349c56e8d5/core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py#L263-L271)

Current `_train()` then uses `rounds < 1`:

- [`lifelong_learning.py:352–369`](https://github.com/kubeedge/ianvs/blob/37a9c60a9747af0cfe3170f84249bc349c56e8d5/core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py#L352-L369)

Therefore, the first standard-mode call receives `rounds == 1` and sets:

```text
HAS_COMPLETED_INITIAL_TRAINING=True
```

The Sedna consumer uses this value to choose `update()` when it reads `"true"`:

- [`sedna/core/lifelong_learning/lifelong_learning.py:241–249`](https://github.com/kubeedge/sedna/blob/2c30e569eedadf6a89c22f3ec6d6310a61c3b66b/lib/sedna/core/lifelong_learning/lifelong_learning.py#L241-L249)

## Executed mode matrix

The CPU-only probe called the real Ianvs `_train()` method from each locked revision. A fake job read `HAS_COMPLETED_INITIAL_TRAINING` through the real Sedna `Context` during `train()`.

| Mode | Round | Desired | `main@37a9c60` | `PR #573@96522f6` |
|---|---:|---|---|---|
| `no-inference` | 0 | `False` | `False` | `False` |
| `no-inference` | 1 | `True` | `True` | `True` |
| `hard-example-mining` | 0 | `False` | `False` | `False` |
| `hard-example-mining` | 1 | `True` | `True` | `True` |
| standard | 1 | `False` | **`True` — incorrect** | **`False` — correct** |
| standard | 2 | `True` | `True` | `True` |

Selected output:

```text
CASE mode=standard round=1 desired=False actual=True matches_desired=False
RESULT=PASS (main reproduces incorrect standard-mode first-round state)
EXIT_CODE=0

CASE mode=standard round=1 desired=False actual=False matches_desired=True
RESULT=PASS (PR #573 matches the desired state matrix for all tested modes)
EXIT_CODE=0
```

> **

<img width="1800" height="1006" alt="Image" src="https://github.com/user-attachments/assets/7df109ea-06a0-4c02-a447-b08ccdd9916e" />

**

## Technical conclusion

The current main behavior is incorrect for the first standard-mode call. The mode-aware implementation in PR #573 produces the desired value for all six tested mode/round combinations and preserves the existing first-round behavior of `no-inference` and `hard-example-mining`.

A regression test should retain this exact matrix rather than testing only `rounds <= 1`, because that simpler condition would incorrectly classify `r = 1` as initial training in the two modes that start at `r = 0`.

## Verification boundary

Executed:

- real Ianvs `_train()` from the locked main and PR #573 revisions;
- real Sedna `Context` lookup;
- all six mode/round combinations above;
- Python 3.11.9, exit code `0` for both runs.

Not executed:

- full GPU/model training;
- external datasets;
- end-to-end benchmark metrics;
- CI or production behavior.

The evidence confirms the state transition itself; it does not claim that every Lifelong Example has completed a full benchmark with the patch.

**Comment by suhaan-24** [comment]
**Confirming the flag logic, plus a second off-by-one in the sibling branch that a fix here should probably account for.**

### Reproduction

On `examples/cityscapes-synthia/lifelong_learning_bench/curb-detection`
(`sedna 0.6.0.1`, Python 3.9.6, macOS/Apple M4, CPU+MPS), the standard path fails inside
sedna's knowledge-base update:

```
[INFO] Download last task index from /home/data/tmp/cloud_task/index.pkl to index.pkl
Exception: index.pkl not exists!
  sedna/algorithms/unseen_task_detection/unseen_sample_re_recognition/
    base_unseen_sample_re_recognition.py:51
```

### The chain

- `sedna/core/lifelong_learning/lifelong_learning.py:241` — `train()` branches on
  `HAS_COMPLETED_INITIAL_TRAINING`. `"true"` routes to `update()`, which **requires** a
  pre-existing KB index. Anything else performs real initial training and needs no KB.
- `core/.../lifelong_learning/lifelong_learning.py` — `_train()` sets the flag to `'False'`
  only when its `rounds` argument is `< 1`.
- the default branch loops `for r in range(1, rounds + 1)`, so the first call passes `r == 1`.
  `1 < 1` is false, the flag is set `'True'`, `update()` runs, and it demands a KB that
  nothing has created yet.

So on the default path the initial-training branch is unreachable, and the KB is required
before anything is able to create it.

### Corroborating detail: the sibling branch is off by one the other way

The `no-inference` branch at the same file loops `for r in range(rounds + 1)` — starting at
`r = 0`, which *does* trigger initial training correctly. That asymmetry is what convinced me
this is an off-by-one rather than intended behaviour.

But that branch has its own boundary bug in the opposite direction: it indexes
`dataset_files[r]` for `r` in `0..rounds`, while
`_split_dataset(splitting_dataset_times=rounds)` returns exactly `rounds` pairs
(`dataset.py: split_dataset(times=rounds)`). It therefore raises `IndexError` at `r == rounds`.
`hard-example-mining` shares the same indexing contract and the same shortfall.

Practical consequence: `mode: "no-inference"` is **not** a usable workaround for this issue —
it trades this failure for an `IndexError`. Worth knowing before recommending it, and worth
deciding whether a fix here should cover both branches so the two do not drift further apart.

**Verification boundary:** executed on `cityscapes-synthia/.../curb-detection` only, CPU/MPS,
no GPU, no gated dataset. After correcting the first-round flag, the run proceeded through
initial training and ultimately produced a leaderboard within 0.00062 of the accuracy
published in that Example's README.

**Comment by suhaan-24** [comment]
Adding a sequencing observation: this defect is the blocker that remains once the module-representation
mismatch is cleared, which makes it more urgent than its individual severity suggests.

## Evidence from running a different fix

I checked out PR #645 — which addresses the `module.py` instance-vs-dict mismatch, not this issue —
into an isolated worktree and ran
`examples/cityscapes-synthia/lifelong_learning_bench/curb-detection` against it, with only the
Example's own config repaired (stale paths corrected, `train_url`→`train_index`, dataset path
repointed). No other change.

That branch's own target defect cleared: the log contains **zero** occurrences of
`'TaskDefinitionByOrigin' object has no attribute 'get'`, and execution reached
`Model loaded successfully`. It then stopped here:

```
[INFO] lifelong_learning.py(316) - Download last task index from /tmp/cloud_task/index.pkl to index.pkl
  File ".../paradigm/lifelong_learning/lifelong_learning.py", line 369, in _train
    res = job.train(train_dataset)
  File ".../sedna/core/lifelong_learning/lifelong_learning.py", line 246, in train
    return self.update(train_data,
  File ".../sedna/core/lifelong_learning/lifelong_learning.py", line 319, in update
Exception: index.pkl not exists!
```

The `return self.update(train_data, ...)` frame is the direct confirmation: sedna took the
`update()` branch on the **first** training call, which is exactly the condition described in this
issue.

## Why this matters for ordering

Anyone repairing a lifelong-learning Example will hit the config layer first, then the module
representation, then this. Each blocker is invisible until the previous one clears, so this issue
tends to be discovered last and can look like a lower priority than it is. In practice it is the
final Core blocker between a repaired Example and a completed training round — nothing downstream
of it is reachable while it stands.

Worth noting for whoever picks it up: after this clears, the next failure on this Example is not a
Core defect but an environment one — sedna's unseen-sample service derives its directory from
`DATA_PATH_PREFIX` and needs `UNSEEN_SAVE_URL` set independently. That is not a reason to delay
this fix, only a note that a green run needs both.

## Verification boundary

Executed on `cityscapes-synthia/lifelong_learning_bench/curb-detection`, Apple M4 / macOS,
Python 3.9.6, sedna 0.6.0.1, CPU+MPS, no GPU and no gated dataset. **I did not check out or run
the branch that fixes this issue** — the evidence above establishes that the defect is reachable
and blocking after other fixes are applied, not that any particular diff clears it.
